File: android/src/main/java/com/rngooglemapsplus/MapCircleBuilder.kt
package com.rngooglemapsplus

import android.graphics.Color
import com.facebook.react.uimanager.PixelUtil.dpToPx
import com.google.android.gms.maps.model.Circle
import com.google.android.gms.maps.model.CircleOptions
import com.rngooglemapsplus.extensions.toColor
import com.rngooglemapsplus.extensions.toLatLng

class MapCircleBuilder {
  fun build(circle: RNCircle): CircleOptions =
    CircleOptions().apply {
      center(circle.center.toLatLng())
      radius(circle.radius)
      circle.strokeWidth?.let { strokeWidth(it.dpToPx()) }
      circle.strokeColor?.let { strokeColor(it.toColor()) }
      circle.fillColor?.let { fillColor(it.toColor()) }
      circle.pressable?.let { clickable(it) }
      circle.zIndex?.let { zIndex(it.toFloat()) }
    }

  fun update(
    prev: RNCircle,
    next: RNCircle,
    circle: Circle,
  ) {
    if (prev.center.latitude != next.center.latitude ||
      prev.center.longitude != next.center.longitude
    ) {
      circle.center = next.center.toLatLng()
    }

    if (prev.radius != next.radius) {
      circle.radius = next.radius
    }

    if (prev.strokeWidth != next.strokeWidth) {
      circle.strokeWidth = next.strokeWidth?.dpToPx() ?: 1f
    }

    if (prev.strokeColor != next.strokeColor) {
      circle.strokeColor = next.strokeColor?.toColor() ?: Color.BLACK
    }

    if (prev.fillColor != next.fillColor) {
      circle.fillColor = next.fillColor?.toColor() ?: Color.TRANSPARENT
    }

    if (prev.pressable != next.pressable) {
      circle.isClickable = next.pressable ?: false
    }

    if (prev.zIndex != next.zIndex) {
      circle.zIndex = next.zIndex?.toFloat() ?: 0f
    }
  }
}

File: ios/MapCircleBuilder.swift
import GoogleMaps

final class MapCircleBuilder {
  func build(_ c: RNCircle) -> GMSCircle {
    let circle = GMSCircle()
    circle.position = c.center.toCLLocationCoordinate2D()
    circle.radius = c.radius
    c.fillColor.map { circle.fillColor = $0.toUIColor() }
    c.strokeColor.map { circle.strokeColor = $0.toUIColor() }
    c.strokeWidth.map { circle.strokeWidth = CGFloat($0) }
    c.pressable.map { circle.isTappable = $0 }
    c.zIndex.map { circle.zIndex = Int32($0) }

    return circle
  }

  func update(_ prev: RNCircle, _ next: RNCircle, _ c: GMSCircle) {
    onMain {
      if !prev.centerEquals(next) {
        c.position = next.center.toCLLocationCoordinate2D()
      }

      if prev.radius != next.radius {
        c.radius = next.radius
      }

      if prev.fillColor != next.fillColor {
        c.fillColor = next.fillColor?.toUIColor() ?? .clear
      }

      if prev.strokeColor != next.strokeColor {
        c.strokeColor = next.strokeColor?.toUIColor() ?? .black
      }

      if prev.strokeWidth != next.strokeWidth {
        c.strokeWidth = CGFloat(next.strokeWidth ?? 1.0)
      }

      if prev.pressable != next.pressable {
        c.isTappable = next.pressable ?? false
      }

      if prev.zIndex != next.zIndex {
        c.zIndex = Int32(next.zIndex ?? 0)
      }
    }
  }
}
